yjit_codegen.c

```diff
@@ -1664,6 +1659,26 @@ guard_two_fixnums(ctx_t* ctx, uint8_t* side_exit)
     val_type_t arg1_type = ctx_get_opnd_type(ctx, OPND_STACK(0));
     val_type_t arg0_type = ctx_get_opnd_type(ctx, OPND_STACK(1));
 
+    if (arg0_type.is_heap || arg1_type.is_heap) {
+        jmp_ptr(cb, side_exit);
+        return;
+    }
+
+    if (arg0_type.type != ETYPE_FIXNUM && arg0_type.type != ETYPE_UNKNOWN) {
+        jmp_ptr(cb, side_exit);
+        return;
+    }
+
+    if (arg1_type.type != ETYPE_FIXNUM && arg1_type.type != ETYPE_UNKNOWN) {
+        jmp_ptr(cb, side_exit);
+        return;
+    }
```

Comment:
IMO it should be invalid to call `guard_two_fixnums` in this situation, but this changeset was already getting a bit long, so I'd like to address that in a separate PR.